### ISSUE TITLE
Automate failure and cancellation testing

### ISSUE THREAD

**Issue by onderkalaci** [citusdata/citus#2001]
We've had a chance to discuss this with @marcocitus and @DimCitus. I think it is fair to say that all of us agreed that the proposed approach below which is simple and quite useful for our purposes.

### Problem

Citus is a distributed database and failures are a fact in distributed systems. Thus, handling failures (and similarly cancellations) properly is very crucial. Citus is already good at it, and we always do a very comprehensive failure and cancellation testing before each release. However, as we decide to have shorter release cycles, the failure and cancellation testing starts to be the bottleneck of the release testing in terms of the time/effort we put into it. Also, there is always the risk of missing bugs with manual testing. 

Automating the failure testing would be a big win in terms of both the quality of the product and the time gain we'd get per release.

(Note that we had the similar thoughts around the concurrency testing, and have already automated most of them via #1464, #1600, #1798 and many more PRs).

### Goal

Provide the basic infrastructure for adding automated failure and cancellation testing in the regression test suits, namely the isolation testing.

Note that we don't want to test failures at random points, instead, we'd want to be able to test specific points. 

Random failures would (a) generate nondeterministic results (b) not cover all the cases we'd want to test. Instead, we'd be able to test specific cases such as:

- Fail worker and/or coordinator just before the `BEGIN` is sent
-  Fail worker and/or coordinator just after `BEGIN` is sent
-  Fail worker and/or coordinator just before the coordinator reads the respond to `BEGIN`
-  Fail worker and/or coordinator just before the `INSERT` is sent
-  And, as you can imagine, many more such cases.

Also, this infrastructure should be able to cancel/terminate backends on both the coordinator and the workers given that we'd want to test both.

### Proposal
Isolation tester seems to be a very convenient way of implementing such thing. I'd like to bucket the required infrastructure into four:

**1. Add ability to terminate and/or cancel backends via distributed transaction id**
	
PostgreSQL provides APIs to terminate (`pg_terminate_backend(pid)`) and cancel backends (`pg_cancel_backend(pid)`). However, they are based on the `pids` of the processes. On the coordinator, we don't know the `pids` of the worker processes that a transaction on the coordinator initiated on the worker nodes. Thus, it'd be convenient if we add the following UDFs based on the distributed transaction ids (actually, distributed transaction ids contain unnecessary fields for this purpose, so it should be fine to implement it with only distributed transaction number):

**`bool citus_terminate_backends(distributedTransactionNumber, nodehost, nodeport)`**: Terminates all the backends on the given node with the given distributed transaction number. Note that, it should also work for coordinator node.
	
**`bool citus_cancel_backends(distributedTransactionNumber, nodehost, nodeport)`**: Cancels all the backends on the given node with the given distributed transaction number. Note that, it should also work for coordinator node.

**2. Add a new session level variable into Citus (and API to set it)**

In Citus core code, add a new session level shared variable (i.e., equivalent of a session-level GUC variable) `char *citus_stop_at`. This variable will be set via a UDF call such as:

```SQL
   SELECT citus_stop_at(text);
```
And, the UDF would simply set the variable to the given value. Later, we'll use this variable in the 4th item below.

**3. Add a new UDF to acquire a lock on the hash of a given text**

The aim is to be able to be able to block some specific session(s) by acquiring a lock on the given key (or more precisely acquire advisory locks on the hash of the given text)

```SQL
SELECT citus_block_at('text')
```

**4. Add a new C function/macro on Citus source code**

This macro would acquire an advisory on the hash of the given text if (a) Assert build mode (b) the session level variable defined in the second item is set to the input value.	
```C
	STOP_AT_IF_VARIABLE_SET('text');
```

Once all of the above infrastructures are implemented, we could add calls in Citus core to stop at any given point via `STOP_AT_IF_VARIABLE_SET` macro and write isolation tests such that they are blocked on the specified points and terminate/cancel backends at those points.

**Example isolation test:**
```SQL
session "control-session"

/* add all the points you'd want to block Citus */
step "control-init"
{
	SELECT citus_block_at('pre_begin');
	SELECT citus_block_at('post_begin');
	SELECT citus_block_at('pre_commit');
	SELECT citus_block_at('post_commit');

	SELECT citus_block_at('pre_router_execution');
	SELECT citus_block_at('during_router_execution');
	SELECT citus_block_at('post_router_execution');
	/* and any more required tests */
}


step "terminate-coordinator"
{
	SELECT citus_terminate_backends(1000, 'localhost', $MASTER_PORT);
}

step "cancel-coordinator"
{
	SELECT citus_cancel_backends(1000, 'localhost', $MASTER_PORT);
}

step "terminate-worker-1"
{
	SELECT citus_terminate_backends(1000, $WORKER_HOST, $WORKER_1_PORT);
}

step "cancel-worker-1"
{
	SELECT citus_cancel_backends(1000, $WORKER_HOST, $WORKER_1_PORT);
}

session "s1"

step "begin-and-assign-tx-id"
{
	BEGIN;
	SELECT assign_distributed_transaction_id(1000, 1000, now());
}

step "insert-stops-at-pre-begin"
{
	SELECT citus_stop_at('pre_begin');
	INSERT INTO table VALUES (1,1);
}


-- below is a sample permutation that could be used for testing failures and cancellations
-- on pre-begin (i.e., just before `BEGIN` is sent to the worker)
permutation "control-init" "begin-and-assign-tx-id" "insert-stops-at-pre-begin" "terminate-coordinator"
permutation "control-init" "begin-and-assign-tx-id" "insert-stops-at-pre-begin" "cancel-coordinator"	
permutation "control-init" "begin-and-assign-tx-id" "insert-stops-at-pre-begin" "terminate-worker-1"
permutation "control-init" "begin-and-assign-tx-id" "insert-stops-at-pre-begin" "cancel-worker-1"

```

### Things to consider before working on the task
* Namings of the UDFs
* Do we want to expose the UDFs to extension (super)-users, or keep it test only
* Do not fill Citus core code with many `STOP_AT_IF_VARIABLE_SET()` calls. We don't want the readability of the codebase to decrease. (Luckily, many of the connection/transaction management stuff is common to every operation. So, hopefully, that wouldn't cause trouble)
* Get rid of/replace the hacky tests that aim to does the same such as `ConflictOnlyWithIsolationTesting ()` [function](https://github.com/citusdata/citus-enterprise/pull/194/files#diff-d39f2511b5e6c9fd73159b73e39b70b3R234)

**Comment by lithp** [comment]
This is awesome! I've had similar thoughts for a while. I even played with systemtap a while back (the Linux answer to Dtrace). Here's a one-liner which tells the process to die right after it returns from `NeedsDistributedPlanning`:

```
sudo stap -x 32391 -g -e 'probe process("/home/brian/Work/pg-10/lib/postgresql/citus.so").function("NeedsDistributedPlanning").return { raise(9) }'
```

systemtap isn't great for us because it requires root (won't work on Travis) and is preeeeety linux-specific. eBPF (and something like bcc) would be cleaner but still require root and are still tied to linux. Sad :cry: 

Another option is GDB; it would let us set breakpoints and kill/block the process anywhere in it's execution without needing to add any macros. It would also let us do fancy things like conditionally kill/block the process (maybe we want to kill the process on the third iteration of some loop).

**Comment by onderkalaci** [comment]
> Another option is GDB; it would let us set breakpoints and kill/block the process anywhere in it's execution without needing to add any macros.

Hey Brain, could we automate such thing in our regression tests (or even on the testing automation framework)?  If we can, it'd be great to compare both approaches.

**Comment by lithp** [comment]
> Hey Brian, could we automate such thing in our regression tests (or even on the testing automation framework)? If we can, it'd be great to compare both approaches.

Doing it inside testing automation would be easier but I think it's also possible from the regression tests. I haven't thought about the right interface to use, I'm not familiar with our failure tests so don't know what they need, but here's _an_ interface:

- `debug_block_at('location')` - attaches gdb to the current process (the one `pg_backend_pid()` would return) and sets a breakpoint at [location](https://sourceware.org/gdb/onlinedocs/gdb/Specify-Location.html#Specify-Location) (the best breakpoints are likely function names, such as `NeedsDistributedPlanning`, line numbers likely won't be stable enough)
- `debug_continue()` - resumes all processes which are stuck at a breakpoint set by `debug_block_at()`

For workers it's a bit more difficult, the process you want to mess with might not yet exist when you call these functions:

- `worker_block_at('location', nodename, nodeport)` - this sets a flag in the coordinator. When this flag is set then Citus will run `debug_block_at` immediately upon opening any new connections to the worker.
- `worker_continue(nodename, nodeport)` - this connects to the given worker and runs `debug_continue()`
- `worker_die_at('location', nodename, nodeport)` - we could also add this convenience function. It's the same as `worker_block_at` but kills the worker process immediately once the breakpoint is triggered.
- I'm not sure what kinds of tests we run but instead of just block and die it's possible to also create a generic UDF which lets us run any gdb script such that we can simulate any kind of behavior.

These functions are probably c-funcs which call `system()`. They call some perl/python script which does the dirty work of invoking gdb.

Here's an example gdb script for implementing `worker_die_at`:

```gdb
break NeedsDistributedPlanning
commands
  kill
end
continue
```

which you can invoke by calling ` sudo gdb --batch -p 7018 --command=commands.gdb`.

If I run it I get this:

```
brian@rhythm:~/Work/pg-10$ bin/psql 
psql (10.0)
Type "help" for help.

brian=# SELECT pg_backend_pid();
 pg_backend_pid 
----------------
           7018
(1 row)

brian=# SELECT 1;
server closed the connection unexpectedly
	This probably means the server terminated abnormally
	before or while processing the request.
The connection to the server was lost. Attempting reset: Failed.
!>
```

Also, I looked into Travis, apparently we have sudo after all! Systemtap might help us not need to write that perl/python script, but the rest of the interface stays the same.

**Comment by jasonmp85** [comment]
> Also, I looked into Travis, apparently we have sudo after all!

We do, but I'm always afraid they're going to take it away from us. Using it has prevented us from moving to their newest images, but they've continued to update the "legacy" ones for some time now.

That said, if they take away `sudo`, I've got some bookmarks of projects that could allow us to use it within a `chroot` or we could create a trivial Docker image for doing our build (and then we'd have `sudo` within the Docker image). So assume we'll always have it, just wanted to provide some context on the history of it (it was default on before, but now it's opt-in for new projects; you have to specify `sudo: true` in your `.travis.yml`).

**Comment by lithp** [comment]
> We do, but I'm always afraid they're going to take it away from us. 

Hmmm, interesting. I did a quick survey and I don't think the kernel is going to give us any help if we want to modify the behavior of another process and we don't own it. This leaves a few options (if we want to ditch sudo):

- non-root is allowed to call `ptrace` on child processes. We could modify postgres to have the postmaster debug backends and forward commands.
- Change the source code (as Onder originally suggested) by manually embedding breakpoints.
- Dynamically intercept calls to Citus functions.

That first one sounds very hard and very not-worth-it. The last one is a bit fanciful but could probably be built without an inordinate amount of effort?

The `LD_PRELOAD` environment variable lets us inject shared libraries, we could use this to intercept any call to a Citus function. We could generate a library which intercepts every call and immediately forwards them to the real library; but can somehow be configured to die or block or do something custom when specific functions are called.

**Comment by jasonmp85** [comment]
@lithp sorry if I added confusion here; I was saying we will probably always have `sudo` and if we don't we can possibly emulate it otherwise (though I suppose I have my reservations around whether `strace`/`stap` will work in a `chroot` or under Docker without elevated container privileges)…

Besides it's impossible to install `.deb` packages from outside the Travis-sanctioned repositories (whose whitelist has not accepted new entries for several years now), so I doubt they'll ever remove `sudo` for that reason alone.

**Comment by jasonmp85** [comment]
I realize I'm a bit late to the party, given how much has been done in #2044, but I'm a little concerned about the difficulty in writing these tests. How exactly do we expect developers to go about this?

**Comment by jasonmp85** [comment]
[`libfiu`](https://blitiri.com.ar/p/libfiu/) looks promising to inject errors into `libpq` calls in the coordinator (which is the only way the coordinator responds to work from the workers, so this makes sense as an approach). Active as recently as September 2017 or so.

[`lfi`](https://github.com/dslab-epfl/lfi) appears neat (there's an Aster slide deck showing its use against PostgreSQL), but its lack of activity makes me thing it would be a bad decision.

**Comment by lithp** [comment]
@jasonmp85 I'm starting to agree with you. [Here's an example test](https://github.com/citusdata/citus/pull/2044/commits/e8caf9fe0cf8cd89afe06356b2ea61b3f3ee63e8), and [here's just the `.spec`](https://github.com/citusdata/citus/blob/e8caf9fe0cf8cd89afe06356b2ea61b3f3ee63e8/src/test/regress/specs/fail_router_no_txn.spec). It could be made easier, both "attach" and the "break-*" steps could be reduced to single function calls, but it's pretty complex! This is a lot of formality for "cut off communication with the worker the first time we call `ConsumeQueryResult` during execution of `UPDATE test_table SET b = 2 WHERE a = 1;`".

So maybe isolationtester isn't the right thing to use for these kinds of tests? But if it isn't, I'm not sure what a better framework would be.

**Comment by lithp** [comment]
`libfiu` looks very interesting! It wouldn't let us inject calls into libpq unless we're testing with a custom build of postgres (it requires code modification unless we're just blocking calls like `read()`, and I think we want more granularity that that). But, as a way of implementing Onder's first proposal it looks great!

And `lfi` looks incredible, they put a lot of work into it! It can only modify return values, so we can't do things like injecting calls to `pg_advisory_lock(1)` sans code modifications, but it looks great for testing failure conditions, we have a few code paths which don't have coverage because our black-box tests can't expose them.

**Comment by jasonmp85** [comment]
> libfiu looks very interesting! It wouldn't let us inject calls into libpq unless we're testing with a custom build of postgres (it requires code modification unless we're just blocking calls like read(), and I think we want more granularity that that). But, as a way of implementing Onder's first proposal it looks great!

Why? Isn't `libpq` dynamically linked to us? We'd just codegen a bunch of pass-through methods based on the interfaces (similar to what they did for their bundled POSIX interposer) and `LD_PRELOAD` our copy somehow, no?

**Comment by jasonmp85** [comment]
(Note, we could also codegen an interposed library of Citus methods and do the same to our own extern methods if we desire).

**Comment by jasonmp85** [comment]
(BTW there are some implicit failure tests in [this suite](https://github.com/citusdata/citus/blob/991000efc9fc5ac36a2aec99df2ebb4a4dc0d46c/src/test/regress/sql/multi_modifying_xacts.sql#L278-L292) which I added long ago by abusing `DEFERRED` check constraints… by inserting certain values to certain shards, I could simulate workers failing right after `BEGIN`, or in the middle, or right before `COMMIT`. "Failing" here meant `ERROR`s, but it's still sort of similar to what you're saying.

**Comment by marcocitus** [comment]
> Also, I looked into Travis, apparently we have sudo after all! 

Do you actually need sudo to run gdb on your own processes on Linux? I've never run gdb as superuser.

**Comment by lithp** [comment]
>Do you actually need sudo to run gdb on your own processes on Linux? I've never run gdb as superuser. 

Oh, interesting. [In Ubuntu](https://wiki.ubuntu.com/SecurityTeam/Roadmap/KernelHardening#ptrace_Protection) it's been disallowed since `10.10`. I kind of assumed other flavors would have it as well, I think [it was upstreamed](https://www.kernel.org/doc/Documentation/security/Yama.txt).

**Comment by onderkalaci** [comment]
We've already implement this with #2119 

We're also tracking the remaining test that we have not covered yet with https://github.com/citusdata/citus/issues/2262

So, I'm feeling comfortable enough to close this. If anyone thinks otherwise, feel free to re-open.

(I think for other kinds of failures such as memory errors we should open a new issue(s), this was mostly considering connection errors)

**Comment by onderkalaci** [comment]
Probably accidentally opened again, see https://github.com/citusdata/citus/issues/2001#issuecomment-505975957
